pkg/webservice/state_dump.go

```diff
@@ -0,0 +1,225 @@
+	var period time.Duration
+	var err error
+	var zapField = zap.Duration("defaultStateDumpPeriodSeconds", defaultStateDumpPeriodSeconds)
+
+	if len(vars["period"]) == 0 {
+		log.Logger().Info("using the default period for state dump",
+			zapField)
+		periodFromHeader = defaultPeriod
+	} else if periodFromHeader, err = strconv.Atoi(vars["period"]); err != nil {
+		log.Logger().Warn("illegal value for period, using the default",
+			zapField)
+		periodFromHeader = defaultPeriod
+	}
+
+	if periodFromHeader < 0 {
+		log.Logger().Warn("period value is negative, using the default",
+			zapField)
+		periodFromHeader = defaultPeriod
+	}
+
+	period = time.Duration(periodFromHeader)
+	if err = startBackGroundStateDump(period); err != nil {
+		buildJSONErrorResponse(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+func disablePeriodicStateDump(w http.ResponseWriter, r *http.Request) {
+	if err := stopBackGroundStateDump(); err != nil {
+		buildJSONErrorResponse(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+func doStateDump(w io.Writer) error {
+	stateDump.Lock()
+	defer stateDump.Unlock()
+
+	partitionContext := schedulerContext.GetPartitionMapClone()
+	records := imHistory.GetRecords()
+	zapConfig := log.GetConfig()
+
+	var aggregated = AggregatedStateInfo{
+		Timestamp:          time.Now().Format(time.RFC3339),
+		Partitions:         getPartitionInfoDAO(partitionContext),
+		Applications:       getApplicationsDAO(partitionContext),
+		AppHistory:         getAppHistoryDAO(records),
+		Nodes:              getNodesDAO(partitionContext),
+		NodesUtilization:   getNodesUtilDAO(partitionContext),
+		ClusterInfo:        getClusterDAO(partitionContext),
+		ClusterUtilization: getClustersUtilDAO(partitionContext),
+		ContainerHistory:   getContainerHistoryDAO(records),
+		Queues:             getPartitionDAO(partitionContext),
+		LogLevel:           zapConfig.Level.Level().String(),
+	}
+
+	var prettyJSON []byte
+	var err error
+	prettyJSON, err = json.MarshalIndent(aggregated, "", "  ")
+	if err != nil {
+		return err
+	}
+	if _, err = w.Write(prettyJSON); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func startBackGroundStateDump(period time.Duration) error {
+	startStop.Lock()
+	defer startStop.Unlock()
+
+	if periodicStateDump {
+		var errMsg = "state dump already running"
+		log.Logger().Error(errMsg)
+		return fmt.Errorf(errMsg)
+	}
+
+	file, err := os.OpenFile(stateDumpFilePath, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0666)
+	if err != nil {
+		log.Logger().Error("unable to open/create file",
+			zap.Error(err))
+		return err
+	}
+	abort = make(chan struct{})
+	periodicStateDump = true
+
+	go func() {
+		ticker := time.NewTicker(period * time.Second)
+
+		for {
+			select {
+			case <-abort:
+				ticker.Stop()
+				file.Close()
+				return
+			case <-ticker.C:
+				if err := doStateDump(file); err != nil {
+					log.Logger().Error("state dump failed", zap.Error(err))
+					if err := stopBackGroundStateDump(); err != nil {
```

Comment:
shouldn't we restart it in case of some error? If it failed once for some reason it is sure that it will fail for the next calls as well?

Reply:
This means a file I/O error, which is unlikely to be recoverable. I'd rather stop when that happens (that's most likely a full or failed disk).